File: app/src/main/java/com/disruption/cookcentral/ui/favs/CachedRecipesAdapter.java
package com.disruption.cookcentral.ui.favs;

import android.annotation.SuppressLint;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.disruption.cookcentral.databinding.CachedRecipeItemBinding;
import com.disruption.cookcentral.models.CachedRecipe;

public class CachedRecipesAdapter extends ListAdapter<CachedRecipe, CachedRecipesAdapter.RecipeViewHolder> {

    private static final DiffUtil.ItemCallback<CachedRecipe> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<CachedRecipe>() {
                @Override
                public boolean areItemsTheSame(@NonNull CachedRecipe oldItem, @NonNull CachedRecipe newItem) {
                    return oldItem.getId() == newItem.getId();
                }

                @SuppressLint("DiffUtilEquals")
                @Override
                public boolean areContentsTheSame(@NonNull CachedRecipe oldItem, @NonNull CachedRecipe newItem) {
                    return oldItem == newItem;
                }
            };

    public CachedRecipesAdapter() {
        super(DIFF_CALLBACK);
    }

    @NonNull
    @Override
    public RecipeViewHolder onCreateViewHolder(@NonNull ViewGroup parent,
                                               int viewType) {
        CachedRecipeItemBinding binding =
                CachedRecipeItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        return new RecipeViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull RecipeViewHolder holder, final int position) {
        CachedRecipe pastry = getItem(position);
        holder.bind(pastry);
    }

    public CachedRecipe getCachedRecipeAtPosition(int position) {
        return getItem(position);
    }

    class RecipeViewHolder extends RecyclerView.ViewHolder {

        private CachedRecipeItemBinding mRecipeBinding;

        RecipeViewHolder(CachedRecipeItemBinding binding) {
            super(binding.getRoot());
            mRecipeBinding = binding;
        }

        private void bind(CachedRecipe cachedRecipe) {
            mRecipeBinding.setRecipe(cachedRecipe);
        }
    }
}

File: app/src/main/java/com/disruption/cookcentral/ui/favs/CachedRecipesAdapter.kt
package com.disruption.cookcentral.ui.favs

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.disruption.cookcentral.databinding.CachedRecipeItemBinding
import com.disruption.cookcentral.models.CachedRecipe

class CachedRecipesAdapter : ListAdapter<CachedRecipe, CachedRecipesAdapter.RecipeViewHolder>(DIFF_CALLBACK) {
    override fun onCreateViewHolder(parent: ViewGroup,
                                    viewType: Int): RecipeViewHolder {
        val binding = CachedRecipeItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        return RecipeViewHolder(binding)
    }

    override fun onBindViewHolder(holder: RecipeViewHolder, position: Int) {
        val pastry = getItem(position)
        holder.bind(pastry)
    }

    fun getCachedRecipeAtPosition(position: Int): CachedRecipe? {
        return getItem(position)
    }

    inner class RecipeViewHolder(private val mRecipeBinding: CachedRecipeItemBinding) : RecyclerView.ViewHolder(mRecipeBinding.root) {
        fun bind(cachedRecipe: CachedRecipe?) {
            mRecipeBinding.recipe = cachedRecipe
        }

    }

    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<CachedRecipe> = object : DiffUtil.ItemCallback<CachedRecipe>() {
            override fun areItemsTheSame(oldItem: CachedRecipe, newItem: CachedRecipe): Boolean {
                return oldItem.id == newItem.id
            }

            @SuppressLint("DiffUtilEquals")
            override fun areContentsTheSame(oldItem: CachedRecipe, newItem: CachedRecipe): Boolean {
                return oldItem === newItem
            }
        }
    }
}